### PR TITLE
Setup wizard: Mode selection screen 

### LINKED ISSUE #4706: Wizard step: Choosing mode
## Feature description

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Display standard, reader, and transitional modes with information explaining each
* Add ability to choose a mode
* Next button and back button

## Implementation brief

* For this step we will create a shell of the React context provider that will handle the site scan. Currently the shell will only need to provide the recommended mode.
* This recommendation will be passed in to the mode choice screen.

## QA testing instructions

1. Pull the branch and run `npm run build` (or switch to the branch if using QA tester plugin)
2. Navigate to the setup wizard
3. Move to the "Template Modes" screen
4. Make a selection and navigate to the last screen of the wizard to force data save
5. Reload the page and verify your selection persisted
6. Verify the reader themes selection screen is skipped if you don't have reader mode selected

**Note**: For testing purposes, simulate the existence of AMP plugin issues by adding `&amp-plugin-issues=1` to the URL. Simulate the existence of AMP theme issues by adding `&amp-theme-issues=1` to the URL.

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Summary

Fixes #4706

This creates the main framework for the Setup Wizard's mode selection screen. Currently it looks like this:

Screen Shot 2020-06-18 at 3 38 03 PM

This screen serves as a bridge between the theme/plugin scan and the final screens. While it lays out some pieces for those screens, those screens are coming later, so this intermediate step is not fully complete.

### Main changes

- Create the layout and styles for the main components on the screen (shown above)
- Creates the function that will generate all the pieces of copy and the mode recommendations for the 8 different combinations of `hasThemeIssues`, `hasPluginIssues`, and `userIsTechnical`. 
- Adds the React state logic needed to skip the Reader Themes screen if the reader mode is not selected. While working on this I refactored the Reader screen a bit. Previously we were fetching reader theme data at the start of the app in all circumstances. Now we'll only fetch the reader themes when the user has selected reader mode. 

## QA

See QA steps on #4706 

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by westonruter** [comment]
While this may come later, I feel like the collapsed sections should have some meaningful information in the collapsed state, namely:

image

If all the sections are collapsed, then it will just show Details and Compatibility repeatedly for each mode. I think it would be better to eliminate the “Compatibility” heading entirely, to make that red/green/blue box as a whole an expandable section (moving the arrow inside of it). The duplicated “Not recommended for your site” can then appear inside this section. 

image

The “Details” heading can probably be replaced with a short one-line description of the mode, with then the ability to expand for more information about it. 

cc @amedina @jwold

**Comment by johnwatkins0** [comment]
I like that suggestion. I will give it a try.

**Comment by amedina** [comment]
I like that approach as well. The Details and Compatibility headings can be eliminated without loss of information.

**Comment by jwold** [comment]
What happens with the one line summary we were going to have?

**Comment by johnwatkins0** [comment]
I have pushed an update based on the above and the subsequent Slack discussion:

Screen Shot 2020-06-22 at 9 55 11 PM

**Comment by johnwatkins0** [comment]
@jwold @westonruter We need an interim solution for this for 1.6 since we won't have site scan results. It will mirror what will happen if the site scan fails (which we were planning to account for anyway). Maybe we could use the blue option for every mode _except_ for non-technical users. For non-technical users we can still make reader mode the most recommended option (i.e. the green notice).

We should probably remove the drawer also because we won't have site scan results to display.

**Comment by jwold** [comment]
@johnwatkins0 do you mean something like this for the 1.6 release? (cc @amedina and @westonruter)

PNG image-3B8B64733F53-1

The idea being that we have a simple notice at the top for each mode (when relevant), and hold off on the expanding detail box until 1.7.

Then for 1.7 we'd release the other idea we've been discussing: https://d.pr/i/IkQWmC

**Comment by johnwatkins0** [comment]
@jwold What I have locally right now differs from your sketch in that we still use the large version of the notice component (that will serve as the drawer in 1.7) but it is not expandable, and we show the blue notice for everything. I think we have gone in the direction of the "Details" line being a general explanation of the mode, and the line in the notice being the reason the user may or may not choose that option. In keeping with this, I think it's useful to keep the notice for the "Why?" factor.

**Non-technical users would see:**
Screen Shot 2020-06-23 at 1 10 38 PM

**Technical users would see:**
Screen Shot 2020-06-23 at 1 10 52 PM

**Comment by jwold** [comment]
@johnwatkins0 thanks for the clarification. I agree with your recommendation! :)

**Comment by pierlon** [comment]
@johnwatkins0 you can revert https://github.com/ampproject/amp-wp/pull/4900/commits/1bb82541934cdccc90613ccda17a18ae069d6025. That change should not be necessary, as CSS files should not reside in `assets/js`. Along with the revert, removing the optimization config imported from the Gutenberg Webpack config should return the original behavior of the file being put in `assets/css`:

```diff
--- webpack.config.js	(revision 1bb82541934cdccc90613ccda17a18ae069d6025)
+++ webpack.config.js	(date 1592969043556)
@@ -50,7 +50,6 @@
 		} ),
 	],
 	optimization: {
-		...defaultConfig.optimization,
 		minimizer: [
 			new TerserPlugin( {
 				parallel: true,
```

---

As to why it is now putting the `amp-setup` CSS file in `assets/js`:

Due to the `amp-setup` entry being over the default Webpack size limit (244 KiB), Webpack will optimize the asset according to the optimization config supplied, which is currently set to optimize CSS files that match `style.css` and put them in the default output folder (`assets/js`). This configuation was imported from the Gutenberg Webpack config and is not relevant to this project, so it can be removed.

File: assets/src/setup/components/amp-notice/index.js
Comment (pierlon) [inline assets/src/setup/components/amp-notice/index.js]: Small nitpick:

Having this all in one line makes this a bit hard to read at first glance. Perhaps having the args on separate lines would help:

```suggestion
	const noticeIcon = useMemo(
		() => icon ? icon : getDefaultNoticeIcon( type ),
		[ icon, type ],
	);
```

File: assets/src/setup/components/amp-notice/index.js
Comment (pierlon) [inline assets/src/setup/components/amp-notice/index.js]:
```suggestion
	size: PropTypes.oneOf( [ NOTICE_SIZE_LARGE, NOTICE_SIZE_SMALL ] ).isRequired,
```

File: assets/src/setup/components/amp-notice/index.js
Comment (pierlon) [inline assets/src/setup/components/amp-notice/index.js]: Is there an instance where a custom icon will be needed?

File: assets/src/setup/pages/template-mode/index.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/index.js:31]: Perhaps we should enforce selecting a technical background before being able to move to the next page.Currently, doing so will result in a intermediate spinner being shown on the next page:

image

File: assets/src/setup/components/navigation-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/navigation-context-provider.js:53]: A similar thing should happen when navigating backwards. If reader mode is not selected, I can navigate backwards to the "Theme selection" page (although the list of themes is not shown):

image

Comment (pierlon) [reply assets/src/setup/components/navigation-context-provider.js:53]: I think the "Theme selection" page should be removed from the navigation and stepper if reader mode is not selected.

File: assets/src/setup/components/options-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/options-context-provider.js]: What's the difference of simply setting the state to `{}`?

File: assets/src/setup/components/options-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/options-context-provider.js]: Any reason why this is no longer required?

File: assets/src/setup/components/site-scan-context-provider.js
Comment (pierlon) [inline assets/src/setup/components/site-scan-context-provider.js]: Are these still needed for development purposes? I suppose they can be removed before merge, or simply replaced when the site scanning feature is ready.

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/get-selection-details.js:34]: Missing param doc for `hasScanResults`.

File: assets/src/setup/pages/template-mode/get-selection-details.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/get-selection-details.js]:
```suggestion
			throw new Error( __( 'A selection text recommendation was not accounted for. ', 'amp' ) + JSON.stringify( args ) );
```

File: assets/src/setup/pages/template-mode/index.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/index.js]:
```suggestion
	// The actual display component should avoid using global context directly. This will facilitate developing and testing the UI using different options.
```

File: assets/src/setup/pages/template-mode/screen-ui.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/screen-ui.js:38]: Missing `@param` for `id`.

File: assets/src/setup/pages/template-mode/test/get-selection-details.js
Comment (pierlon) [inline assets/src/setup/pages/template-mode/test/get-selection-details.js]: Maybe put each import on a new line?

File: tests/e2e/specs/amp-onboarding/index.js
Comment (pierlon) [inline tests/e2e/specs/amp-onboarding/index.js]:
```suggestion
		return page.$eval( 'h1', ( el ) => el.innerText );
```

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/setup/components/amp-notice/index.js
Comment (johnwatkins0) [reply assets/src/setup/components/amp-notice/index.js]: Nah, there was a different icon in one of the earlier comps but it's not there anymore. I'm going to remove the icon prop until it's needed.

File: assets/src/setup/components/navigation-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/navigation-context-provider.js:53]: What we decided for this was that we'd skip "Theme Selection" but allow users to return to it. What I'd missed was updating that text to indicate that the screen is not relevant for users who have not selected reader. I've updated.

File: assets/src/setup/components/options-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/options-context-provider.js]: Yeah, that wasn't necessary.

File: assets/src/setup/components/site-scan-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/site-scan-context-provider.js]: I've updated this comment to explain more what's going on. Essentially we need to keep everything here, but right now this component is passing `null` for theme issues and plugin issues, imitating what will happen in the future if the site scan is unsuccessful for some reason.

File: assets/src/setup/pages/template-mode/index.js
Comment (johnwatkins0) [reply assets/src/setup/pages/template-mode/index.js:31]: Thanks, that was easy to miss. I've updated.

File: assets/src/setup/components/options-context-provider.js
Comment (johnwatkins0) [reply assets/src/setup/components/options-context-provider.js]: This was for unit testing purposes, but I've revised the approach; rather than changing the prop here (it should be required), in the test I've just mocked an OptionsContextProvider that doesn't fetch.

**Comment by johnwatkins0** [comment]
Thanks for the review @pierlon . I have addressed all the feedback. @westonruter Let me know if you see anything else, noting the update I've made to handle the absence of site scan results in 1.6 (see screenshot above).

File: assets/src/setup/pages/choose-reader-theme/index.js
Comment (westonruter) [inline assets/src/setup/pages/choose-reader-theme/index.js:51]:
```suggestion
				{ __( 'This screen is only relevant to sites that use Reader mode. Please continue.', 'amp' ) }
```

File: assets/src/setup/pages/template-mode/screen-ui.js
Comment (westonruter) [inline assets/src/setup/pages/template-mode/screen-ui.js:79]:
```suggestion
					<a href="http://amp-wp.org" target="_blank">
```

File: assets/src/setup/pages/template-mode/screen-ui.js
Comment (westonruter) [inline assets/src/setup/pages/template-mode/screen-ui.js:93]:
```suggestion
							<a href="http://amp-wp.org" target="_blank">
```

**Review by westonruter — APPROVED** [review]
Good evolution on the screens.

**Pull request merged**
